Src/Extern/HYPRE/AMReX_HypreABecLap3.cpp

```diff
@@ -0,0 +1,503 @@
+  const int comp = 0;
+  const Real* dx = geom.CellSize();
+  for (int n=0; n< BL_SPACEDIM; ++n) {
+    for (MFIter mfi(acoefs); mfi.isValid(); ++mfi) {
+      int i = mfi.index();
+
+      const Box& bx = mfi.validbox();
+
+      // Our default will be that the face of this grid is
+      // either touching another grid
+      // across an interior boundary or a periodic boundary.
+      // We will test for the other
+      // cases below.
+      {
+        // Define the type of boundary conditions to be Dirichlet
+        // (even for periodic)
+        bd.setBoundCond(Orientation(n, Orientation::low), i,
+                        comp, LO_DIRICHLET);
+        bd.setBoundCond(Orientation(n, Orientation::high), i,
+                        comp, LO_DIRICHLET);
+
+        // Set the boundary conditions to the
+        // cell centers outside the domain
+        bd.setBoundLoc(Orientation(n, Orientation::low), i, 0.5*dx[n]);
+        bd.setBoundLoc(Orientation(n, Orientation::high), i, 0.5*dx[n]);
+      }
+
+      // Now test to see if we should override
+      // the above with Dirichlet or Neumann physical bc's
+      if(bc_type != LinOpBCType::interior) {
+        int ibnd = static_cast<int>(bc_type);
+        // either LO_DIRICHLET or LO_NEUMANN
+
+        // We are on the low side of the
+        // domain in coordinate direction n
+
+        if (bx.smallEnd(n) == geom.Domain().smallEnd(n)) {
+          // Set the boundary conditions to
+          // live exactly on the faces of the domain
+          bd.setBoundLoc(Orientation(n, Orientation::low), i, 0.0);
+
+          // Set the Dirichlet/Neumann boundary values
+          bd.setValue(Orientation(n, Orientation::low), i, bc_value);
+
+          // Define the type of boundary conditions
+          bd.setBoundCond(Orientation(n, Orientation::low),
+                          i, comp, ibnd);
+        }
+
+        // We are on the high side of the
+        // domain in coordinate direction n
+        if (bx.bigEnd(n) == geom.Domain().bigEnd(n)) {
+          // Set the boundary conditions to
+          // live exactly on the faces of the domain
+          bd.setBoundLoc(Orientation(n, Orientation::high), i, 0.0);
+
+          // Set the Dirichlet/Neumann boundary values
+          bd.setValue(Orientation(n, Orientation::high), i, bc_value);
+
+          // Define the type of boundary conditions
+          bd.setBoundCond(Orientation(n, Orientation::high),
+                          i, comp, ibnd);
+        }
+      }
+    }
+  }
+}
+
+void HypreABecLap3::loadMatrix() {
+  static_assert(BL_SPACEDIM > 1, "HypreABecLap2: 1D not supported");
+
+  const int size = 2 * BL_SPACEDIM + 1;
+  const int bho = 0;
+  const Real* dx = geom.CellSize();
+
+  for (MFIter mfi(acoefs); mfi.isValid(); ++mfi) {
+    int i = mfi.index();
+    const Box &reg = mfi.validbox();
+
+    int volume = reg.numPts();
+
+    // build matrix interior
+
+    amrex_hmac_ij(BL_TO_FORTRAN(acoefs[mfi]), ARLIM(reg.loVect()),
```

Comment:
amrex_hmac_ij and other Fortran routines call functions like `HYPRE_HYPRE_IJMatrixAddToValues`.  These are C functions, right?  I am worried about portability issues of calling them from Fortran.  Also `integer(kind=8)` does not necessarily mean an 8-bytes integer.

I think we do the following,

    IArrayBox cols, rows, ncols;
    FArrayBox values;
    for (MFIter mfi(acoefs); ....) {
         // .....
        cols.resize(reg);  rows.resize(reg); ....; values.resize(reg);
        amrex_hmac_ij(...., cols.dataPtr(), rows.dataPtr(), ncols.dataPtr(), values.dataPtr(),...);
        HYPRE_IJMatrixAddtoValuse(A, nrows.dataPtr(), ncols.dataPtr(), ....);
    }

    subroutine hmac_ij(....., cols, ...)
        integer :: cols(*)
    end subroutine hmac_ij

Reply:
I see later you call `HYPRE_IJMatrixAddToValues` inside loops in Fortran.  So my approach above won't work in that case.  An alternative approach is pass `&A` as type(c_ptr) by value to Fortran.  In Fortran, we call our own C function that in turn call Hypre.

    amrex_hmmat_ij(..., &A, ....);

    void amrex_hmmat_ij(..., void*, ...);  // in _F.H 

    interface
      void our_own_c_function_that_wraps_HYPRE_IJMatrixAddToValues(A,...) bind(c)
          type(c_ptr), value :: A
    end interface

     subroutine hmac_ij(..., Amat, ...) bind(c)
        type(c_ptr), value :: Amat
        ! ...
        call our_own_c_function_that_wraps_HYPRE_IJMatrixAddToValues(Amax,...)

     void our_own_c_function_that_wraps_HYPRE_IJMatrixAddToValues(double* A, ...)  // Is there Hypre_real that's typedefed double?